crates/worker/src/purgatory.rs

```diff
@@ -64,57 +91,83 @@ impl Drop for Inner {
 }
 
 impl Inner {
-    fn new(handle: NonNull<JoinHandle<()>>, state: State) -> Self {
+    fn new(
+        handle: NonNull<JoinHandle<()>>,
+        state: State,
+        db_client: DynamoDBClient,
+        s3_client: S3Client,
+    ) -> Self {
         Self {
-            handle,
             state,
+            s3_client,
+            db_client,
+            handle,
             _marker: PhantomData,
         }
     }
 
-    pub fn purge(&mut self) {
-        todo!()
+    fn add_record(&mut self, id: Uuid, result: TaskResult) {
+        pub const DURATION_TO_PURGE: u64 = 60 * 5; // 5 minutes
+        let to_purge_timestampe = timestamp() + DURATION_TO_PURGE;
+
+        self.state.task_status.insert(id, Status::Done(result));
+        self.state
+            .expiration_timestamps
+            .push((id, to_purge_timestampe));
     }
 
-    // TODO: replace with Self::purge
-    // pub async fn supervisor(
-    //     db_client: DynamoDBClient,
-    //     expiration_timestamps: Arc<Mutex<Vec<(Uuid, Timestamp)>>>,
-    // ) {
-    //     loop {
-    //         let now = timestamp();
-    //
-    //         let to_delete = {
-    //             let mut to_delete = vec![];
-    //             let mut expiration_timestamps = expiration_timestamps.lock().await;
-    //             expiration_timestamps.retain(|&(uuid, expiration)| {
-    //                 if expiration < now {
-    //                     to_delete.push(uuid);
-    //                     false
-    //                 } else {
-    //                     true
-    //                 }
-    //             });
-    //
-    //             to_delete
-    //         };
-    //
-    //         for uuid in to_delete {
-    //             db_client.delete_item(uuid.to_string()).await;
-    //         }
-    //
-    //         sleep(Duration::from_millis(2000)).await;
-    //     }
-    // }
-}
+    pub async fn purge(&mut self) {
+        let now = timestamp();
+        for (id, timestamp) in self.state.expiration_timestamps.iter() {
+            if *timestamp > now {
+                break;
+            }
+
+            let status = if let Some(status) = self.state.task_status.get(id) {
+                status
+            } else {
+                warn!("Inconsistency! ID present vector but not in status map");
+                continue;
+            };
+
+            match status {
+                Status::InProgress => warn!("Item compiling for too long!"),
```

Comment:
We can add some force-stop if compilation goes beyond some time

Reply:
Right now local purgatory may contain only `Status::Done`. It is global purger that also utilizes that function for purging.

Pending - is not something that worker could support since if its in worker its not longer pending. 
InProgress - is complicated to reason about, since it could mean whether we failed to reach `add_record` or compilation takes too long. Global purger shalln't do anything in that case, for local one the course of action if obscure. So I left it empty for now